#! /bin/sh
patch -p1 -l -f $* < $0
exit $?

From: Kevin Backhouse <kevinbackhouse@github.com>
Date: Wed, 3 Sep 2025 14:36:54 +0100
Subject: [PATCH 7/7] Check for duplicate entries

(cherry picked from commit 4ce27cc826bf90cc8dbbd8a8c87bd913cccd7ec0)

CVE-2025-52885
---
 poppler/StructTreeRoot.cc | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/poppler/StructTreeRoot.cc b/poppler/StructTreeRoot.cc
index 99ddbfe5..f7a27943 100644
--- a/poppler/StructTreeRoot.cc
+++ b/poppler/StructTreeRoot.cc
@@ -137,6 +137,10 @@ void StructTreeRoot::parseNumberTreeNode(Dict *node)
                 }
                 int keyVal = key.getInt();
                 std::vector<Parent> &vec = parentTree[keyVal];
+                if (!vec.empty()) {
+                    error(errSyntaxError, -1, "Nums item at position {0:d} is a duplicate entry for key {1:d}", i, keyVal);
+                    continue;
+                }

                 Object valueArray = nums.arrayGet(i + 1);
                 if (valueArray.isArray()) {
--
2.47.2
